### ISSUE TITLE
Custom Dataset

### ISSUE THREAD

**Issue by bmmo** [petko-nikolov/semseg.pytorch#4]
Hi,
I am using a custom dataset for training but it is not clear what are the ideal crop sizes of each model, namely UNet, FCDenseNet-xxx and DeeplabV3?
Are any transforms done for training/inference?

**Comment by petko-nikolov** [comment]
Hi bmmo, 

The crop size depends on many factors and likely trying few different values might be the best. Some factors/thoughts:

-  your compute capabilities could be limiting, e.g. if you have a GPU with 8 GB of RAM and train with batch size of ~4, 512x512 is close to the max crop size for UNet and DeeplabV3+, for FCDenseNet it's even smaller
- FCDenseNet-xxx trains well on CamVid and a crop size of 224x224, but in their paper they don't provide results on larger image size datasets. In my experience it's really hard to make it work with higher resolution images and on a dataset with large number of classes (>20).
- the crop size / image size ratio shouldn't be too small as you may lose important context near the edges of the image. If your images are higres what you can try is downsampling them to a 512 (smaller side size) and then use a similarly sized crops. If you then need higher res masks you can upsample by interpolation the output mask from the model. 

_Are any transforms done for training/inference?_

Yes, you can take a look at the transforms defined for PascalVoc [here](https://github.com/petko-nikolov/pysemseg/blob/master/pysemseg/datasets/pascal_voc/pascal.py#L146). For your custom dataset, you can define similar transformer class and pass it to `pysemseg-train` as in the repo description. You can change the crop size and tune other augmentations there. With the `mode` parameter you can control what transforms are applied during `train/val/test`

This repo is a good example how you can use the library: https://github.com/petko-nikolov/spacenet

**Comment by bmmo** [comment]
Hi @petko-nikolov,
Thanks for the reply. After investing a lot of time on FC-DenseNet with 512x512 crops, I can confirm that  it won't provide any good results even after 800+ epochs. Very unstable.
You're also right about the crop/image size ratio. Thanks for the feedback!

**Comment by bmmo** [comment]
I've written my own dataset script (customdataset.py on my sandbox folder, class CustomDataSet inheriting from pysemseg.SegmentationDataset). But now I can't import it from the pysemseg-train script. 

I'm using a custom config.yaml, but when I specify the dataset name that is on my sandbox folder, the pysemseg-train script can't find it. Is there a way around this or am I missing something?

**Comment by petko-nikolov** [comment]
Ensure that your dataset class can be imported from the folder from which you are executing the script. As an example, if this executes successfully

```
from path.to.my.dataset import CustomDataSet
```

then the way you should specify it in the config will be

```
...
dataset: path.to.my.dataset.CustomDataSet
...
```

Ensure that all folders along the path to CustomDataSet have an `__init__.py` file.

Please let me know whether this worked for you, I am happy to take a look if you specify more details.

**Comment by bmmo** [comment]
Many thanks for the reply. I was doing that, but my problem was that I had a dataset.py script along with a dataset/ folder with the same name and that was messing the python import system. 
Thanks for the help!
